source/core/slang-riff.cpp

```diff
@@ -0,0 +1,99 @@
+#include "slang-riff.h"
+
+#include "../../slang-com-helper.h"
+
+namespace Slang
+{
+
+/* static */int64_t RiffUtil::calcChunkTotalSize(const RiffChunk& chunk)
+{
+    int64_t size = chunk.m_size + sizeof(RiffChunk);
+    return (size + 3) & ~int64_t(3);
+}
+
+/* static */SlangResult RiffUtil::skip(const RiffChunk& chunk, Stream* stream, int64_t* remainingBytesInOut)
+{
+    int64_t chunkSize = calcChunkTotalSize(chunk);
+    if (remainingBytesInOut)
+    {
+        *remainingBytesInOut -= chunkSize;
+    }
+
+    // Skip the payload (we don't need to skip the Chunk because that was already read
+    stream->Seek(SeekOrigin::Current, chunkSize - sizeof(RiffChunk));
+    return SLANG_OK;
+}
+
+
+/* static */SlangResult RiffUtil::readChunk(Stream* stream, RiffChunk& outChunk)
+{
+    try
+    {
+        stream->Read(&outChunk, sizeof(RiffChunk));
+    }
+    catch (IOException&)
+    {
+    	return SLANG_FAIL;
+    }
+
+    // TODO(JS): Could handle endianness issues here...
+
+    return SLANG_OK;
+}
+
+
+/* static */SlangResult RiffUtil::writeData(uint32_t riffType, const void* data, size_t size, Stream* out)
+{
+    SLANG_ASSERT(uint64_t(size) <= uint64_t(0xfffffffff));
+
+    // TODO(JS): Could handle endianness here
+    RiffChunk chunk;
+    chunk.m_type = riffType;
+    chunk.m_size = uint32_t(size);
+
+    try
+    {
+        out->Write(&chunk, sizeof(chunk));
+        out->Write(data, size);
+        size_t remaining = size & 3;
+        if (remaining)
+        {
+            uint8_t end[4] = { 0, 0, 0, 0};
+            out->Write(end, 4 - remaining);
+        }
+    }
+    catch (IOException&)
+    {
+        return SLANG_FAIL;
+    }
+
+    return SLANG_OK;
+}
+
+
+/* static */SlangResult RiffUtil::readData(Stream* stream, RiffChunk& outChunk, List<uint8_t>& data)
+{
```

Comment:
This is a longer-term discussion, but one reason why it might be good to migrate away from the RIFF-based container toward something more ELF-like is that object/executable file formats are better designed for being `mmap`-able in cases where their bits can be used directly (which is in some ways the whole point of the relative-pointer encoding you are using). In particular, an ELF-style sectioned format lets you ensure that the in-file offset of a section is consistent with its in-memory alignment requirement (whereas RIFF has a much less strict alignment guarantee; it looks like we are doing 4-byte alignment here, but RIFF only enforces 2-byte IIUC).

There are certainly cases where each style is appropriate, and I'm not looking to change things right away, but I want to make note of this stuff.

Reply:
Oh, and it is a good thing to start having this container-format stuff pulled out of the IR serialization and into its own infrastructure so we'll be able to apply it more broadly.